system-tests/ft-suite/src/test/kotlin/com/hotels/styx/resiliency/OriginResourcesSpec.kt

```diff
@@ -0,0 +1,161 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.resiliency
+
+import com.github.tomakehurst.wiremock.client.WireMock
+import com.hotels.styx.api.HttpHeaderNames.HOST
+import com.hotels.styx.api.HttpRequest.get
+import com.hotels.styx.api.HttpResponseStatus.OK
+import com.hotels.styx.client.StyxHttpClient
+import com.hotels.styx.server.HttpConnectorConfig
+import com.hotels.styx.servers.MockOriginServer
+import com.hotels.styx.support.StyxServerProvider
+import com.hotels.styx.support.proxyHttpHostHeader
+import com.hotels.styx.support.wait
+import io.kotlintest.Spec
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.StringSpec
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.Files.copy
+import java.nio.file.Path
+import java.nio.file.StandardCopyOption.REPLACE_EXISTING
+
+class OriginResourcesSpec : StringSpec() {
+    val count = 20
+
+    val mockServer = MockOriginServer.create("", "", 0, HttpConnectorConfig(0))
+            .start()
+            .stub(WireMock.get(WireMock.urlMatching("/.*")), WireMock.aResponse()
+                    .withStatus(200)
+                    .withBody("mock-server-01"))
+
+
+    init {
+        "Cleans up threads after use" {
+            "---" + (1..count)
+                    .map { appDeclaration("aaa-$it") }
+                    .joinToString("\n")
+                    .let { writeConfig(styxOriginsFile, it) }
+
+            // Wait for styx to pick up the configuration
+            Thread.sleep(2000)
+
+            (1..count).forEach {
+                client.send(get("/aaa-$it").header(HOST, styxServer().proxyHttpHostHeader()).build())
+                        .wait(debug = false)
+                        .status() shouldBe OK
+            }
+
+            val threadCountBefore = threadCount("Styx-Client-Worker")
+
+            "---" + (1..count)
+                    .map { appDeclaration("bbb-$it") }
+                    .joinToString("\n")
+                    .let { writeConfig(styxOriginsFile, it) }
+
+            // Wait for styx to pick up the new configuration
+            Thread.sleep(2000)
```

Comment:
We should use `io.kotlintest.eventually` instead of a long sleep.

Reply:
Not sure if I'd use here. The following step is not really a verification step, but rather a step to ensure that new client worker threads are being created.

Reply:
That's not the point.
1. Eventually can be given a short interval (e.g. 100 ms) will exit once the job is done, sleep(2000) will never spend less than 2 seconds. The more we use sleeps, the more we unnecessarily extend the duration of the build, which we do not want.
2. Sleep will exit after 2 seconds even if the job is not done. This makes the test behaviour nondeterministic. If some other process on the machine causes a delay, the test will not work properly and the whole build has to be run again.

Reply:
I know the benefits of `eventual` @kvosper. I just wanted to keep the code less cluttered. But I'll think about something.

Reply:
Okay. I did some improvements on this test. Including eventual. Have another look please.